File: Python/Data Structure/Linked List/linked_list.py
class Node:
    def __init__(self, value) -> None:
        self.value = value
        self.next = None
    
class LinkedList:
    def __init__(self) -> None:
        self.head = None
        self.tail = self.head
        self.length = 0
    
    def is_empty(self):
        if not self.head:
            return True
        return False
    
    def pre_append(self, value) -> None:
        new_node = Node(value)

        if self.is_empty():
            self.head = new_node
            self.tail = self.head
            self.length += 1
            return None

        new_node.next = self.head
        self.head = new_node
        self.length += 1
        return None
    
    def append(self, value) -> None:
        new_node = Node(value)

        if self.is_empty():
            self.head = new_node
            self.tail = self.head
            self.length += 1
            return None

        self.tail.next = new_node
        self.tail = new_node
        self.length += 1
        return None
    
    def insert(self, index, value):
        if index <= 0:
            raise IndexError("Index out of range")
        
        if self.is_empty():
            self.append(value)
            return None

        if index == 1:
            self.pre_append(value)
            return None

        new_node = Node(value)
        previous_node = self.traverse(index-1)
        new_node.next = previous_node.next
        previous_node.next = new_node
        self.length += 1
        return None
    
    def delete(self, index):
        if index <= 0:
            return IndexError("Index out of range")
        
        if index == 1:
            self.head = self.head.next
            self.length -= 1
            return None
        
        previous_node = self.traverse(index-1)
        unwanted_node = previous_node.next
        previous_node.next = unwanted_node.next
        self.length -= 1

        return None
    
    def traverse(self, index):
        if self.is_empty():
            return None
        current_node = self.head

        for _ in range(1, index):
            if not self.head:
                return None
            current_node = current_node.next
        return current_node
        

    def search(self, index):
        if self.is_empty():
            return None
        current_node = self.head

        for _ in range(1, index):
            if not self.head:
                return None
            current_node = current_node.next
        return current_node.value
    
    def print_list(self) -> list:
        li = []

        if self.is_empty():
            return li
        
        current_node = self.head

        while current_node:
            li.append(current_node.value)
            current_node = current_node.next
        
        return li

    def get_length(self):
        return self.length

linked_list = LinkedList()

linked_list.append('C')
linked_list.append('JavaScript')
linked_list.append('Python')
linked_list.append('Haskell')
linked_list.append('Clojure')
linked_list.append('Odin')

print(linked_list.print_list())
print(linked_list.get_length())

linked_list.pre_append('Assembly')

print(linked_list.print_list())

print(linked_list.insert(8, 'S'))
print(linked_list.print_list())

linked_list.delete(7)

print(linked_list.print_list())

File: JavaScript/Data Structure/Linked List/linkedList.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.tail = this.head;
        this.len = 0;
    }

    isEmpty() {
        return !this.head;
    }

    preAppend(value) {
        const newNode = new Node(value);

        if (this.isEmpty()) {
            this.addToEmptyList(newNode);
            return;
        }
        newNode.next = this.head;
        this.head = newNode;
        this.len++;
        return;
    }

    append(value) {
        const newNode = new Node(value);

        if (this.isEmpty()) {
            this.addToEmptyList(newNode);
            return;
        }
        this.tail.next = newNode;
        this.tail = newNode;
        this.len++;
        return;
    }

    insert(index, value) {
        if (index < 0) {
            return "Index out of range";
        }
        if (index === 1) {
            this.preAppend(value);
            return;
        }
        const newNode = new Node(value);
        let previousNode = this.traverse(index - 1);

        newNode.next = previousNode.next;
        previousNode.next = newNode;
        this.len++;

        if(!newNode.next) {
            this.tail = newNode;
        }

        return;
    }

    pop() {
        const item = this.head.value;
        this.head = this.head.next;
        this.len--;
        return item;
    }

    delete(index) {
        if (index < 0 || index > this.len) {
            return "Index out of range";
        }
        if (index === 1) {
            this.pop();
            return;
        }
        let previousNode = this.traverse(index - 1);
        let currentNode = previousNode.next;

        previousNode.next = currentNode.next;
        this.len--;

        if(!previousNode.next) {
            this.tail = previousNode;
        }

        return;
    }

    addToEmptyList(newNode) {
        this.head = newNode;
        this.tail = this.head;
        this.len++;
    }

    traverse(index) {
        let currentNode = this.head;

        for (let i = 1; i < index; i++) {
            if (!currentNode) {
                return;
            }
            currentNode = currentNode.next;
        }
        return currentNode;
    }

    print() {
        if (this.isEmpty()) {
            return "The List is empty";
        }
        const array = [];
        let currentNode = this.head;

        while (currentNode) {
            array.push(currentNode.value);
            currentNode = currentNode.next;
        }
        return array;
    }
}

const linkedList = new LinkedList();

linkedList.append('C');
linkedList.append('JavaScript');
linkedList.append('Python');
linkedList.append('Julia');
linkedList.append('Haskell');
linkedList.append('Clojure');
linkedList.append('Odin');

console.log(linkedList.print());

linkedList.preAppend('Assembly');
linkedList.preAppend('Machine Code');

linkedList.insert(10, 'x64 Assembly');

console.log(linkedList.print());

linkedList.delete(10);

console.log(linkedList.print());
